// @catalyst-team/poly-sdk@0.3.0/dist/services/trading-service.js
/**
 * TradingService
 *
 * Trading service using official @polymarket/clob-client.
 *
 * Provides:
 * - Order creation (limit, market)
 * - Order management (cancel, query)
 * - Rewards tracking
 * - Balance management
 *
 * Note: Market data methods have been moved to MarketService.
 */
import { ClobClient, Side as ClobSide, OrderType as ClobOrderType, } from '@polymarket/clob-client';
import { Wallet } from 'ethers';
import { ApiType } from '../core/rate-limiter.js';
import { PolymarketError, ErrorCode } from '../core/errors.js';
// Chain IDs
export const POLYGON_MAINNET = 137;
export const POLYGON_AMOY = 80002;
// CLOB Host
const CLOB_HOST = 'https://clob.polymarket.com';
// ============================================================================
// TradingService Implementation
// ============================================================================
export class TradingService {
    rateLimiter;
    cache;
    config;
    clobClient = null;
    wallet;
    chainId;
    credentials = null;
    initialized = false;
    tickSizeCache = new Map();
    negRiskCache = new Map();
    constructor(rateLimiter, cache, config) {
        this.rateLimiter = rateLimiter;
        this.cache = cache;
        this.config = config;
        this.wallet = new Wallet(config.privateKey);
        this.chainId = (config.chainId || POLYGON_MAINNET);
        this.credentials = config.credentials || null;
    }
    // ============================================================================
    // Initialization
    // ============================================================================
    async initialize() {
        if (this.initialized)
            return;
        // Create CLOB client with L1 auth (wallet)
        this.clobClient = new ClobClient(CLOB_HOST, this.chainId, this.wallet);
        // Get or create API credentials
        if (!this.credentials) {
            const creds = await this.clobClient.createOrDeriveApiKey();
            this.credentials = {
                key: creds.key,
                secret: creds.secret,
                passphrase: creds.passphrase,
            };
        }
        // Re-initialize with L2 auth (credentials)
        this.clobClient = new ClobClient(CLOB_HOST, this.chainId, this.wallet, {
            key: this.credentials.key,
            secret: this.credentials.secret,
            passphrase: this.credentials.passphrase,
        });
        this.initialized = true;
    }
    async ensureInitialized() {
        if (!this.initialized || !this.clobClient) {
            await this.initialize();
        }
        return this.clobClient;
    }
    // ============================================================================
    // Trading Helpers
    // ============================================================================
    /**
     * Get tick size for a token
     */
    async getTickSize(tokenId) {
        if (this.tickSizeCache.has(tokenId)) {
            return this.tickSizeCache.get(tokenId);
        }
        const client = await this.ensureInitialized();
        const tickSize = await client.getTickSize(tokenId);
        this.tickSizeCache.set(tokenId, tickSize);
        return tickSize;
    }
    /**
     * Check if token is neg risk
     */
    async isNegRisk(tokenId) {
        if (this.negRiskCache.has(tokenId)) {
            return this.negRiskCache.get(tokenId);
        }
        const client = await this.ensureInitialized();
        const negRisk = await client.getNegRisk(tokenId);
        this.negRiskCache.set(tokenId, negRisk);
        return negRisk;
    }
    // ============================================================================
    // Order Creation
    // ============================================================================
    /**
     * Create and post a limit order
     */
    async createLimitOrder(params) {
        const client = await this.ensureInitialized();
        return this.rateLimiter.execute(ApiType.CLOB_API, async () => {
            try {
                const [tickSize, negRisk] = await Promise.all([
                    this.getTickSize(params.tokenId),
                    this.isNegRisk(params.tokenId),
                ]);
                const orderType = params.orderType === 'GTD' ? ClobOrderType.GTD : ClobOrderType.GTC;
                const result = await client.createAndPostOrder({
                    tokenID: params.tokenId,
                    side: params.side === 'BUY' ? ClobSide.BUY : ClobSide.SELL,
                    price: params.price,
                    size: params.size,
                    expiration: params.expiration || 0,
                }, { tickSize, negRisk }, orderType);
                const success = result.success === true ||
                    (result.success !== false &&
                        ((result.orderID !== undefined && result.orderID !== '') ||
                            (result.transactionsHashes !== undefined && result.transactionsHashes.length > 0)));
                return {
                    success,
                    orderId: result.orderID,
                    orderIds: result.orderIDs,
                    errorMsg: result.errorMsg,
                    transactionHashes: result.transactionsHashes,
                };
            }
            catch (error) {
                return {
                    success: false,
                    errorMsg: `Order failed: ${error instanceof Error ? error.message : String(error)}`,
                };
            }
        });
    }
    /**
     * Create and post a market order
     */
    async createMarketOrder(params) {
        const client = await this.ensureInitialized();
        return this.rateLimiter.execute(ApiType.CLOB_API, async () => {
            try {
                const [tickSize, negRisk] = await Promise.all([
                    this.getTickSize(params.tokenId),
                    this.isNegRisk(params.tokenId),
                ]);
                const orderType = params.orderType === 'FAK' ? ClobOrderType.FAK : ClobOrderType.FOK;
                const result = await client.createAndPostMarketOrder({
                    tokenID: params.tokenId,
                    side: params.side === 'BUY' ? ClobSide.BUY : ClobSide.SELL,
                    amount: params.amount,
                    price: params.price,
                }, { tickSize, negRisk }, orderType);
                const success = result.success === true ||
                    (result.success !== false &&
                        ((result.orderID !== undefined && result.orderID !== '') ||
                            (result.transactionsHashes !== undefined && result.transactionsHashes.length > 0)));
                return {
                    success,
                    orderId: result.orderID,
                    orderIds: result.orderIDs,
                    errorMsg: result.errorMsg,
                    transactionHashes: result.transactionsHashes,
                };
            }
            catch (error) {
                return {
                    success: false,
                    errorMsg: `Market order failed: ${error instanceof Error ? error.message : String(error)}`,
                };
            }
        });
    }
    // ============================================================================
    // Order Management
    // ============================================================================
    async cancelOrder(orderId) {
        const client = await this.ensureInitialized();
        return this.rateLimiter.execute(ApiType.CLOB_API, async () => {
            try {
                const result = await client.cancelOrder({ orderID: orderId });
                return { success: result.canceled ?? false, orderId };
            }
            catch (error) {
                throw new PolymarketError(ErrorCode.ORDER_FAILED, `Cancel failed: ${error instanceof Error ? error.message : String(error)}`);
            }
        });
    }
    async cancelOrders(orderIds) {
        const client = await this.ensureInitialized();
        return this.rateLimiter.execute(ApiType.CLOB_API, async () => {
            try {
                const result = await client.cancelOrders(orderIds);
                return { success: result.canceled ?? false, orderIds };
            }
            catch (error) {
                throw new PolymarketError(ErrorCode.ORDER_FAILED, `Cancel orders failed: ${error instanceof Error ? error.message : String(error)}`);
            }
        });
    }
    async cancelAllOrders() {
        const client = await this.ensureInitialized();
        return this.rateLimiter.execute(ApiType.CLOB_API, async () => {
            try {
                const result = await client.cancelAll();
                return { success: result.canceled ?? false };
            }
            catch (error) {
                throw new PolymarketError(ErrorCode.ORDER_FAILED, `Cancel all failed: ${error instanceof Error ? error.message : String(error)}`);
            }
        });
    }
    async getOpenOrders(marketId) {
        const client = await this.ensureInitialized();
        return this.rateLimiter.execute(ApiType.CLOB_API, async () => {
            const orders = await client.getOpenOrders(marketId ? { market: marketId } : undefined);
            return orders.map((o) => {
                const originalSize = Number(o.original_size) || 0;
                const filledSize = Number(o.size_matched) || 0;
                return {
                    id: o.id,
                    status: o.status,
                    tokenId: o.asset_id,
                    side: o.side.toUpperCase(),
                    price: Number(o.price) || 0,
                    originalSize,
                    filledSize,
                    remainingSize: originalSize - filledSize,
                    associateTrades: o.associate_trades || [],
                    createdAt: o.created_at,
                };
            });
        });
    }
    async getTrades(marketId) {
        const client = await this.ensureInitialized();
        return this.rateLimiter.execute(ApiType.CLOB_API, async () => {
            const trades = await client.getTrades(marketId ? { market: marketId } : undefined);
            return trades.map((t) => ({
                id: t.id,
                tokenId: t.asset_id,
                side: t.side,
                price: Number(t.price) || 0,
                size: Number(t.size) || 0,
                fee: Number(t.fee_rate_bps) || 0,
                timestamp: Number(t.match_time) || Date.now(),
            }));
        });
    }
    // ============================================================================
    // Rewards
    // ============================================================================
    async isOrderScoring(orderId) {
        const client = await this.ensureInitialized();
        return this.rateLimiter.execute(ApiType.CLOB_API, async () => {
            const result = await client.isOrderScoring({ order_id: orderId });
            return result.scoring;
        });
    }
    async areOrdersScoring(orderIds) {
        const client = await this.ensureInitialized();
        return this.rateLimiter.execute(ApiType.CLOB_API, async () => {
            return await client.areOrdersScoring({ orderIds });
        });
    }
    async getEarningsForDay(date) {
        const client = await this.ensureInitialized();
        return this.rateLimiter.execute(ApiType.CLOB_API, async () => {
            const earnings = await client.getEarningsForUserForDay(date);
            return earnings.map(e => ({
                date: e.date,
                conditionId: e.condition_id,
                assetAddress: e.asset_address,
                makerAddress: e.maker_address,
                earnings: e.earnings,
                assetRate: e.asset_rate,
            }));
        });
    }
    async getCurrentRewards() {
        const client = await this.ensureInitialized();
        return this.rateLimiter.execute(ApiType.CLOB_API, async () => {
            const rewards = await client.getCurrentRewards();
            return rewards.map(r => ({
                conditionId: r.condition_id,
                question: r.question,
                marketSlug: r.market_slug,
                eventSlug: r.event_slug,
                rewardsMaxSpread: r.rewards_max_spread,
                rewardsMinSize: r.rewards_min_size,
                tokens: r.tokens.map(t => ({
                    tokenId: t.token_id,
                    outcome: t.outcome,
                    price: t.price,
                })),
                rewardsConfig: r.rewards_config.map(c => ({
                    assetAddress: c.asset_address,
                    startDate: c.start_date,
                    endDate: c.end_date,
                    ratePerDay: c.rate_per_day,
                    totalRewards: c.total_rewards,
                })),
            }));
        });
    }
    // ============================================================================
    // Balance & Allowance
    // ============================================================================
    async getBalanceAllowance(assetType, tokenId) {
        const client = await this.ensureInitialized();
        return this.rateLimiter.execute(ApiType.CLOB_API, async () => {
            const result = await client.getBalanceAllowance({
                asset_type: assetType,
                token_id: tokenId,
            });
            return { balance: result.balance, allowance: result.allowance };
        });
    }
    async updateBalanceAllowance(assetType, tokenId) {
        const client = await this.ensureInitialized();
        return this.rateLimiter.execute(ApiType.CLOB_API, async () => {
            await client.updateBalanceAllowance({
                asset_type: assetType,
                token_id: tokenId,
            });
        });
    }
    // ============================================================================
    // Account Info
    // ============================================================================
    getAddress() {
        return this.wallet.address;
    }
    getWallet() {
        return this.wallet;
    }
    getCredentials() {
        return this.credentials;
    }
    isInitialized() {
        return this.initialized;
    }
    getClobClient() {
        return this.clobClient;
    }
}
