// @catalyst-team/cache@0.2.0/dist/adapters/memory-adapter.js
export class MemoryAdapter {
    store = new Map();
    cleanupInterval;
    defaultTTL;
    constructor(options) {
        this.defaultTTL = options?.defaultTTL || 60; // Default 60 seconds
        // Periodically clean up expired entries to prevent memory leaks
        const cleanupMs = options?.cleanupIntervalMs || 60000; // Default 1 minute
        this.cleanupInterval = setInterval(() => {
            this.cleanup();
        }, cleanupMs);
    }
    async get(key) {
        const entry = this.store.get(key);
        if (!entry)
            return null;
        if (Date.now() > entry.expiresAt) {
            this.store.delete(key);
            return null;
        }
        return entry.value;
    }
    async set(key, value, ttl) {
        const ttlSeconds = ttl ?? this.defaultTTL;
        this.store.set(key, {
            value,
            expiresAt: Date.now() + ttlSeconds * 1000,
        });
    }
    async del(key) {
        this.store.delete(key);
    }
    async exists(key) {
        const entry = this.store.get(key);
        if (!entry)
            return false;
        if (Date.now() > entry.expiresAt) {
            this.store.delete(key);
            return false;
        }
        return true;
    }
    async clear() {
        this.store.clear();
    }
    /**
     * Get cache statistics
     */
    size() {
        return this.store.size;
    }
    /**
     * Clean up expired entries
     */
    cleanup() {
        const now = Date.now();
        for (const [key, entry] of this.store.entries()) {
            if (now > entry.expiresAt) {
                this.store.delete(key);
            }
        }
    }
    /**
     * Destroy the adapter and stop cleanup interval
     */
    destroy() {
        if (this.cleanupInterval) {
            clearInterval(this.cleanupInterval);
            this.cleanupInterval = undefined;
        }
        this.store.clear();
    }
}
