// @catalyst-team/poly-sdk@0.4.0/dist/src/services/dip-arb-types.js
/**
 * Dip Arbitrage Service Types
 *
 * 暴跌套利服务类型定义
 *
 * 策略原理：
 * 在 Polymarket 的 BTC/ETH/SOL/XRP UP/DOWN 短期市场中：
 *
 * 1. 每个市场有一个 "price to beat"（开盘时的 Chainlink 价格）
 * 2. 结算规则：
 *    - UP 赢：结束时价格 >= price to beat
 *    - DOWN 赢：结束时价格 < price to beat
 *
 * 3. 套利流程：
 *    - Leg1：检测暴跌 → 买入暴跌侧
 *    - Leg2：等待对冲条件 → 买入另一侧
 *    - 利润：总成本 < $1 时获得无风险利润
 */
/**
 * 默认配置
 */
export const DEFAULT_DIP_ARB_CONFIG = {
    shares: 20,
    sumTarget: 0.92, // ✅ 放宽到 0.92 提高 Leg2 成交率 (8%+ 利润)
    dipThreshold: 0.15,
    windowMinutes: 2,
    slidingWindowMs: 3000, // 3秒滑动窗口 - 核心参数！
    maxSlippage: 0.02,
    minProfitRate: 0.03,
    leg2TimeoutSeconds: 180, // ✅ 缩短到 3 分钟，更快退出未对冲仓位
    enableSurge: true,
    surgeThreshold: 0.15,
    autoMerge: true,
    autoExecute: false,
    executionCooldown: 3000,
    splitOrders: 1, // ✅ 默认不拆分，避免份额误差
    orderIntervalMs: 500, // 拆分订单间隔 500ms
    debug: false,
};
/**
 * 默认自动轮换配置
 */
export const DEFAULT_AUTO_ROTATE_CONFIG = {
    enabled: false,
    underlyings: ['BTC'],
    duration: '15m',
    preloadMinutes: 2,
    autoSettle: true,
    settleStrategy: 'redeem',
    redeemWaitMinutes: 5,
    redeemRetryIntervalSeconds: 30,
};
// ============= Helper Functions =============
/**
 * 创建初始统计
 */
export function createDipArbInitialStats() {
    return {
        startTime: Date.now(),
        runningTimeMs: 0,
        roundsMonitored: 0,
        roundsCompleted: 0,
        roundsSuccessful: 0,
        roundsExpired: 0,
        signalsDetected: 0,
        leg1Filled: 0,
        leg2Filled: 0,
        totalSpent: 0,
        totalProfit: 0,
        avgProfitRate: 0,
    };
}
/**
 * 创建新轮次状态
 */
export function createDipArbRoundState(roundId, priceToBeat, upPrice, downPrice, durationMinutes = 15) {
    const now = Date.now();
    return {
        roundId,
        startTime: now,
        endTime: now + durationMinutes * 60 * 1000,
        priceToBeat,
        openPrices: {
            up: upPrice,
            down: downPrice,
        },
        phase: 'waiting',
    };
}
/**
 * 计算利润率
 */
export function calculateDipArbProfitRate(totalCost) {
    if (totalCost >= 1 || totalCost <= 0)
        return 0;
    return (1 - totalCost) / totalCost;
}
/**
 * 计算基于底层资产价格变化的"真实"胜率
 *
 * @param currentPrice - 当前价格
 * @param priceToBeat - 开盘价格
 * @returns UP 的真实胜率估计 (0-1)
 */
export function estimateUpWinRate(currentPrice, priceToBeat) {
    if (priceToBeat <= 0)
        return 0.5;
    const priceChange = (currentPrice - priceToBeat) / priceToBeat;
    // 简单模型：价格变化 1% 对应胜率变化约 10%
    const sensitivity = 10;
    const winRateShift = priceChange * sensitivity;
    // 限制在 [0.05, 0.95] 范围内
    return Math.max(0.05, Math.min(0.95, 0.5 + winRateShift));
}
/**
 * 检测定价偏差
 *
 * @param tokenPrice - token 当前价格（隐含胜率）
 * @param estimatedWinRate - 基于价格估计的真实胜率
 * @returns 偏差程度（正数 = 被低估，负数 = 被高估）
 */
export function detectMispricing(tokenPrice, estimatedWinRate) {
    return estimatedWinRate - tokenPrice;
}
/**
 * 从 slug 解析底层资产
 * e.g., 'btc-updown-15m-1767165300' → 'BTC'
 */
export function parseUnderlyingFromSlug(slug) {
    const lower = slug.toLowerCase();
    if (lower.startsWith('btc'))
        return 'BTC';
    if (lower.startsWith('eth'))
        return 'ETH';
    if (lower.startsWith('sol'))
        return 'SOL';
    if (lower.startsWith('xrp'))
        return 'XRP';
    return 'BTC'; // default
}
/**
 * 从 slug 解析时长
 * e.g., 'btc-updown-15m-1767165300' → 15
 */
export function parseDurationFromSlug(slug) {
    if (slug.includes('-5m-'))
        return 5;
    if (slug.includes('-15m-'))
        return 15;
    return 15; // default
}
/**
 * 类型守卫：检查是否为 Leg1 信号
 */
export function isDipArbLeg1Signal(signal) {
    return signal.type === 'leg1';
}
/**
 * 类型守卫：检查是否为 Leg2 信号
 */
export function isDipArbLeg2Signal(signal) {
    return signal.type === 'leg2';
}
