// @catalyst-team/cache@0.1.0/src/adapters/memory-adapter.ts
/**
 * Memory Cache Adapter (L1)
 *
 * In-memory cache with TTL support.
 * - Fastest access (no I/O)
 * - Lost on process restart
 * - Not shared across instances
 */
import type { CacheAdapter } from '../types.js';

interface CacheEntry<T> {
  value: T;
  expiresAt: number;
}

export class MemoryAdapter implements CacheAdapter {
  private store = new Map<string, CacheEntry<unknown>>();
  private cleanupInterval?: NodeJS.Timeout;
  private readonly defaultTTL: number;

  constructor(options?: { defaultTTL?: number; cleanupIntervalMs?: number }) {
    this.defaultTTL = options?.defaultTTL || 60; // Default 60 seconds

    // Periodically clean up expired entries to prevent memory leaks
    const cleanupMs = options?.cleanupIntervalMs || 60000; // Default 1 minute
    this.cleanupInterval = setInterval(() => {
      this.cleanup();
    }, cleanupMs);
  }

  async get<T>(key: string): Promise<T | null> {
    const entry = this.store.get(key) as CacheEntry<T> | undefined;
    if (!entry) return null;

    if (Date.now() > entry.expiresAt) {
      this.store.delete(key);
      return null;
    }

    return entry.value;
  }

  async set<T>(key: string, value: T, ttl?: number): Promise<void> {
    const ttlSeconds = ttl ?? this.defaultTTL;
    this.store.set(key, {
      value,
      expiresAt: Date.now() + ttlSeconds * 1000,
    });
  }

  async del(key: string): Promise<void> {
    this.store.delete(key);
  }

  async exists(key: string): Promise<boolean> {
    const entry = this.store.get(key);
    if (!entry) return false;

    if (Date.now() > entry.expiresAt) {
      this.store.delete(key);
      return false;
    }

    return true;
  }

  async clear(): Promise<void> {
    this.store.clear();
  }

  /**
   * Get cache statistics
   */
  size(): number {
    return this.store.size;
  }

  /**
   * Clean up expired entries
   */
  private cleanup(): void {
    const now = Date.now();
    for (const [key, entry] of this.store.entries()) {
      if (now > entry.expiresAt) {
        this.store.delete(key);
      }
    }
  }

  /**
   * Destroy the adapter and stop cleanup interval
   */
  destroy(): void {
    if (this.cleanupInterval) {
      clearInterval(this.cleanupInterval);
      this.cleanupInterval = undefined;
    }
    this.store.clear();
  }
}
