// @catalyst-team/cache@0.1.0/src/adapters/redis-adapter.ts
/**
 * Redis Cache Adapter (L2)
 *
 * Redis-backed cache with:
 * - Persistent across process restarts
 * - Shared between multiple instances
 * - Automatic serialization/deserialization
 */
import type Redis from 'ioredis';
import type { CacheAdapter } from '../types.js';

export class RedisAdapter implements CacheAdapter {
  private readonly defaultTTL: number;

  constructor(
    private client: Redis,
    options?: { defaultTTL?: number }
  ) {
    this.defaultTTL = options?.defaultTTL || 300; // Default 5 minutes
  }

  async get<T>(key: string): Promise<T | null> {
    try {
      const value = await this.client.get(key);
      if (value === null) return null;

      // Try to parse as JSON
      try {
        return JSON.parse(value) as T;
      } catch {
        // If parsing fails, return as-is (for string values)
        return value as T;
      }
    } catch (error) {
      console.error('[RedisAdapter] Error getting key:', key, error);
      return null;
    }
  }

  async set<T>(key: string, value: T, ttl?: number): Promise<void> {
    try {
      const ttlSeconds = ttl ?? this.defaultTTL;
      const serialized = typeof value === 'string' ? value : JSON.stringify(value);

      if (ttlSeconds > 0) {
        await this.client.setex(key, ttlSeconds, serialized);
      } else {
        await this.client.set(key, serialized);
      }
    } catch (error) {
      console.error('[RedisAdapter] Error setting key:', key, error);
      throw error;
    }
  }

  async del(key: string): Promise<void> {
    try {
      await this.client.del(key);
    } catch (error) {
      console.error('[RedisAdapter] Error deleting key:', key, error);
      throw error;
    }
  }

  async exists(key: string): Promise<boolean> {
    try {
      const result = await this.client.exists(key);
      return result === 1;
    } catch (error) {
      console.error('[RedisAdapter] Error checking key existence:', key, error);
      return false;
    }
  }

  async clear(): Promise<void> {
    try {
      await this.client.flushdb();
    } catch (error) {
      console.error('[RedisAdapter] Error clearing cache:', error);
      throw error;
    }
  }

  /**
   * Get time-to-live for a key (in seconds)
   */
  async ttl(key: string): Promise<number> {
    try {
      return await this.client.ttl(key);
    } catch (error) {
      console.error('[RedisAdapter] Error getting TTL:', key, error);
      return -2; // Key does not exist
    }
  }

  /**
   * Set expiration time for a key
   */
  async expire(key: string, seconds: number): Promise<boolean> {
    try {
      const result = await this.client.expire(key, seconds);
      return result === 1;
    } catch (error) {
      console.error('[RedisAdapter] Error setting expiration:', key, error);
      return false;
    }
  }

  /**
   * Increment a numeric value
   */
  async incr(key: string): Promise<number> {
    try {
      return await this.client.incr(key);
    } catch (error) {
      console.error('[RedisAdapter] Error incrementing:', key, error);
      throw error;
    }
  }

  /**
   * Decrement a numeric value
   */
  async decr(key: string): Promise<number> {
    try {
      return await this.client.decr(key);
    } catch (error) {
      console.error('[RedisAdapter] Error decrementing:', key, error);
      throw error;
    }
  }

  /**
   * Get multiple keys at once
   */
  async mget<T>(keys: string[]): Promise<(T | null)[]> {
    try {
      const values = await this.client.mget(...keys);
      return values.map((value) => {
        if (value === null) return null;
        try {
          return JSON.parse(value) as T;
        } catch {
          return value as T;
        }
      });
    } catch (error) {
      console.error('[RedisAdapter] Error getting multiple keys:', error);
      return keys.map(() => null);
    }
  }

  /**
   * Set multiple keys at once
   */
  async mset<T>(entries: Array<{ key: string; value: T; ttl?: number }>): Promise<void> {
    try {
      const pipeline = this.client.pipeline();

      for (const { key, value, ttl } of entries) {
        const serialized = typeof value === 'string' ? value : JSON.stringify(value);
        const ttlSeconds = ttl ?? this.defaultTTL;

        if (ttlSeconds > 0) {
          pipeline.setex(key, ttlSeconds, serialized);
        } else {
          pipeline.set(key, serialized);
        }
      }

      await pipeline.exec();
    } catch (error) {
      console.error('[RedisAdapter] Error setting multiple keys:', error);
      throw error;
    }
  }

  /**
   * Delete keys matching a pattern (use with caution!)
   */
  async delPattern(pattern: string): Promise<number> {
    try {
      const keys = await this.client.keys(pattern);
      if (keys.length === 0) return 0;
      return await this.client.del(...keys);
    } catch (error) {
      console.error('[RedisAdapter] Error deleting pattern:', pattern, error);
      throw error;
    }
  }
}
