// @catalyst-team/cache@0.1.0/dist/index.js
/**
 * @catalyst-team/cache
 *
 * Three-layer cache package for Prediction Router V2
 * - L1: Memory cache (in-process)
 * - L2: Redis cache (shared, persistent)
 * - L3: PostgreSQL (handled by application layer)
 */
// Adapters
export { MemoryAdapter } from './adapters/memory-adapter.js';
export { RedisAdapter } from './adapters/redis-adapter.js';
// Core
export { RedisClient } from './redis-client.js';
export { CacheManager } from './cache-manager.js';
// Imports for factory functions
import { MemoryAdapter } from './adapters/memory-adapter.js';
import { RedisAdapter } from './adapters/redis-adapter.js';
import { RedisClient } from './redis-client.js';
import { CacheManager } from './cache-manager.js';
// Convenience factory functions
export function createMemoryCache(options) {
    return new MemoryAdapter(options);
}
export function createRedisCache(options) {
    const { defaultTTL, ...clientOptions } = options || {};
    const client = new RedisClient(clientOptions);
    return new RedisAdapter(client.getClient(), { defaultTTL });
}
export function createCacheManager(options) {
    const l1 = options?.l1 ? createMemoryCache(options.l1) : undefined;
    const l2 = options?.l2 ? createRedisCache(options.l2) : undefined;
    return new CacheManager({ l1, l2 });
}
